File: packages/sdk-python/commons/pagination.py
"""Auto-paginating iterators for cursor-based pagination."""

from __future__ import annotations

from typing import Any, Callable, Dict, Generic, Iterator, List, Optional, TypeVar

from .types import PaginationMeta

T = TypeVar("T")


class CursorPage(Generic[T]):
    """A single page of results with auto-pagination support.

    Iterating over a ``CursorPage`` yields items from the current page, then
    automatically fetches subsequent pages until all results have been consumed.
    """

    def __init__(
        self,
        data: List[T],
        meta: PaginationMeta,
        fetch_next: Optional[Callable[[str], "CursorPage[T]"]] = None,
    ) -> None:
        self.data = data
        self.meta = meta
        self.has_more: bool = meta.get("hasMore", False)
        self._fetch_next = fetch_next

    def __iter__(self) -> Iterator[T]:
        page: Optional[CursorPage[T]] = self
        while page is not None:
            yield from page.data
            cursor = page.meta.get("cursor")
            if page.has_more and cursor and page._fetch_next:
                page = page._fetch_next(cursor)
            else:
                page = None

    def __len__(self) -> int:
        return len(self.data)

    @property
    def total(self) -> Optional[int]:
        return self.meta.get("total")


class AsyncCursorPage(Generic[T]):
    """Async variant of :class:`CursorPage`."""

    def __init__(
        self,
        data: List[T],
        meta: PaginationMeta,
        fetch_next: Optional[Any] = None,
    ) -> None:
        self.data = data
        self.meta = meta
        self.has_more: bool = meta.get("hasMore", False)
        self._fetch_next = fetch_next

    async def __aiter__(self):  # type: ignore[override]
        page: Optional[AsyncCursorPage[T]] = self
        while page is not None:
            for item in page.data:
                yield item
            cursor = page.meta.get("cursor")
            if page.has_more and cursor and page._fetch_next:
                page = await page._fetch_next(cursor)
            else:
                page = None

    def __len__(self) -> int:
        return len(self.data)

    @property
    def total(self) -> Optional[int]:
        return self.meta.get("total")

File: packages/sdk-typescript/src/pagination.ts
import type { PaginationMeta } from './types.js';

export type PageFetcher<T> = (cursor: string | null) => Promise<{ data: T[]; meta: PaginationMeta }>;

export class CursorPage<T> implements AsyncIterable<T> {
  readonly data: T[];
  readonly meta: PaginationMeta;

  private readonly _fetchPage: PageFetcher<T>;

  constructor(data: T[], meta: PaginationMeta, fetchPage: PageFetcher<T>) {
    this.data = data;
    this.meta = meta;
    this._fetchPage = fetchPage;
  }

  get hasMore(): boolean {
    return this.meta.hasMore;
  }

  async nextPage(): Promise<CursorPage<T> | null> {
    if (!this.meta.hasMore || !this.meta.cursor) return null;
    const { data, meta } = await this._fetchPage(this.meta.cursor);
    return new CursorPage(data, meta, this._fetchPage);
  }

  async *[Symbol.asyncIterator](): AsyncIterator<T> {
    // Yield items from current page
    for (const item of this.data) {
      yield item;
    }

    // Fetch subsequent pages
    let nextCursor = this.meta.hasMore ? this.meta.cursor : null;
    while (nextCursor) {
      const { data, meta } = await this._fetchPage(nextCursor);
      for (const item of data) {
        yield item;
      }
      nextCursor = meta.hasMore ? meta.cursor : null;
    }
  }
}
